sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -748,15 +659,8 @@ int gcoap_req_init(coap_pkt_t *pdu, uint8_t *buf, size_t len,
 #endif
 
     if (hdrlen > 0) {
-        /* Reserve some space between the header and payload to write options later */
-        pdu->payload      = buf + coap_get_total_hdr_len(pdu) + strlen(path)
-                                                              + GCOAP_REQ_OPTIONS_BUF;
-        /* Payload length really zero at this point, but we set this to the available
-         * length in the buffer. Allows us to reconstruct buffer length later. */
-        pdu->payload_len  = len - (pdu->payload - buf);
-        pdu->content_type = COAP_FORMAT_NONE;
-
-        memcpy(&pdu->url[0], path, strlen(path));
+        coap_pkt_init(pdu, buf, len - GCOAP_REQ_OPTIONS_BUF, hdrlen);
+        coap_opt_add_string(pdu, COAP_OPT_URI_PATH, path, '/');
```

Comment:
Just for clarification: the `coap_opt_add_x` functions do require to be called in the order of the applied option numbers, correct? So does this mean, that we can not add any option with an opt num < COAP_OPT_URI_PATH (11) after calling `gcoap_req_init()`? Or am I completely of track right now?

Reply:
Yes, you're right that options must be added in order (except see last paragraph). gcoap_req_init() was meant to be a simple way to create a request for the common case where the user does not need to add lower-numbered options. The great thing about the rework of option handling and this PR is that we now have the option to write a message from scratch using coap_build_hdr() and coap_pkt_init(). You could even use gcoap_req_init() as a template for how to do that. 

Actually, now that I look at it, we could allow the path parameter for gcoap_req_init() to be NULL and not write the path option. I'll create a separate PR for that. I like it!

Also, open #9475 sorts options, so in that case options don't need to be added in order. :-)

Reply:
ok, then I understood it correctly. Should we as intermediate step include a `@waring` in the function doc to explicitly state that one should not add any options with opt num < COAP_OPT_URI_PATH when using this function? I'll leave this up to you...